HW4_Sapozhnikov/prototool.py

```diff
@@ -0,0 +1,564 @@
+            'PROTEIN_TO_RNA_COMBINATION': {'AUG'},
+            'PKA_AMINOACIDS': [2.28, 9.21],
+            'MOLECULAR_WEIGHTS': 149},
+    'Phe': {'TO_1': 'F',
+            'PROTEIN_TO_RNA_COMBINATION': {'UUU', 'UUC'},
+            'PKA_AMINOACIDS': [2.20, 9.13],
+            'MOLECULAR_WEIGHTS': 165},
+    'Pro': {'TO_1': 'P',
+            'PROTEIN_TO_RNA_COMBINATION': {'CCU', 'CCC', 'CCA', 'CCG'},
+            'PKA_AMINOACIDS': [1.99, 10.96],
+            'MOLECULAR_WEIGHTS': 115},
+    'Ser': {'TO_1': 'S',
+            'PROTEIN_TO_RNA_COMBINATION': {'UCU', 'UCC', 'UCA', 'UCG'},
+            'PKA_AMINOACIDS': [2.21, 9.15],
+            'MOLECULAR_WEIGHTS': 105},
+    'Thr': {'TO_1': 'T',
+            'PROTEIN_TO_RNA_COMBINATION': {'ACU', 'ACC', 'ACA', 'ACG'},
+            'PKA_AMINOACIDS': [2.11, 9.62],
+            'MOLECULAR_WEIGHTS': 119},
+    'Tyr': {'TO_1': 'W',
+            'PROTEIN_TO_RNA_COMBINATION': {'UAU', 'UAC'},
+            'PKA_AMINOACIDS': [2.20, 9.11, 10.07],
+            'MOLECULAR_WEIGHTS': 181},
+    'Trp': {'TO_1': 'Y',
+            'PROTEIN_TO_RNA_COMBINATION': {'UGG'},
+            'PKA_AMINOACIDS': [2.38, 9.39],
+            'MOLECULAR_WEIGHTS': 204},
+    'Val': {'TO_1': 'V',
+            'PROTEIN_TO_RNA_COMBINATION': {'GUU', 'GUC', 'GUA', 'GUG'},
+            'PKA_AMINOACIDS': [2.32, 9.62],
+            'MOLECULAR_WEIGHTS': 117},
+}
+
+# A dictionary where keys are 1-letter and values are 3-letters codes
+TO_3_DICT = {nested_dict['TO_1']: key for key,
+             nested_dict in AMINOACIDS_DICT.items()}
+
+TRANSCRIBE_DICT: dict = {'A': 'A',
+                         'U': 'T',
+                         'G': 'G',
+                         'C': 'C',
+                         'a': 'a',
+                         'u': 't',
+                         'g': 'g',
+                         'c': 'c'}
+
+
+def is_one_letter(seq: str) -> bool:
+    """
+    Defines whether the sequence is 1 coded.
+
+    Args:
+    - seq - sequence to check
+
+    Returns:
+    - bool
+    """
+    return all(aa.isalpha() and aa.isupper() for aa in seq)
+
+
+def recode(seq: str) -> dict:
+    """
+    Translate 1-letter to 3-letter encoding if 1-letter
+    encoded sequence is given and vice versa.
+
+    Args:
+    - seq - sequence or list of sequences to recode
+
+    Returns:
+    - function_result - a dictionary containing recoded sequences as values
+    for original sequences keys
+    """
+
+    if is_one_letter(seq):
+        # Translate 1-letter to 3-letter coded sequence
+        three_letter_sequence = ""
+        for aa in seq:
+            three_letter_code = TO_3_DICT.get(aa, aa)
+            three_letter_sequence += three_letter_code
+        return three_letter_sequence
+    # Translate 3-letter to 1-letter coded sequence
+    one_letter_sequence = ""
+    for aa in range(0, len(seq), 3):
+        amino_acid = seq[aa:aa+3]
+        one_letter_sequence += AMINOACIDS_DICT[amino_acid]['TO_1']
+    return one_letter_sequence
```

Comment:
Вообще хорошо, но я бы чуть-чуть подправил структуру. У вас по сути вход -> перекодировка -> выход. Но два return в коде получаются как-то логически не равнозначны. Сложно описать, но код функции выглядит "не симметрично" что ли:)
```suggestion
    recoded_sequence = ""
    if is_one_letter(seq):
        for aa in seq:
            three_letter_code = TO_3_DICT.get(aa, aa)
            recoded_sequence += three_letter_code
    else:
        for aa in range(0, len(seq), 3):
            amino_acid = seq[aa:aa+3]
            recoded_sequence += AMINOACIDS_DICT[amino_acid]['TO_1']
    return recoded_sequence 
```
Как-то подобным образом, мб еще можно отрефакторить. 
Также у вас в 2 случае вы шагаете не по `aa`, а по индексу. Это особенно смущает на моменте `aa+3`. Что это значит? 

```python
for aa_idx in range(0, len(seq), 3):
    amino_acid = seq[aa_idx :aa_idx +3]
```